src/middleware/devtools.ts

```diff
@@ -61,107 +74,190 @@ export const devtools =
+          )
+          didWarnAboutDevtools = true
+        }
+        extension = value
+      },
+    })
+
+    let didWarnAboutPrefix = false
+    Object.defineProperty(extension, 'prefix', {
+      get: () => {
+        if (!didWarnAboutPrefix) {
+          console.warn(
+            '[zustand devtools middleware] along with `api.devtools`, `api.devtools.prefix` is deprecated.\n' +
+              'We no longer prefix the actions/names' +
+              devtoolsOptions.name ===
+              undefined
+              ? ', pass the `name` option to create a separate instance of devtools for each store.'
+              : ', because the `name` option already creates a separate instance of devtools for each store.'
+          )
+          didWarnAboutPrefix = true
+        }
+        return '';
+      },
+      set: () => {
+        if (!didWarnAboutPrefix) {
+          console.warn(
+            '[zustand devtools middleware] along with `api.devtools`, `api.devtools.prefix` is deprecated.\n' +
+              'We no longer prefix the actions/names' +
+              devtoolsOptions.name ===
+              undefined
+              ? ', pass the `name` option to create a separate instance of devtools for each store.'
+              : ', because the `name` option already creates a separate instance of devtools for each store.'
+          )
+          didWarnAboutPrefix = true
+        }
+      },
+    })
+
+    let isRecording = true
+    ;(api.setState as NamedSet<S>) = (state, replace, nameOrAction) => {
       set(state, replace)
-      if (!api.dispatch && api.devtools) {
-        api.devtools.send(api.devtools.prefix + (name || 'action'), get())
-      }
+      if (!isRecording) return
+      extension!.send(
+        nameOrAction === undefined
+          ? { type: devtoolsOptions.anonymousActionType ?? 'anonymous' } :
+        typeof nameOrAction === 'string'
+          ? { type: nameOrAction } :
+        nameOrAction,
+        get()
+      )
     }
-    api.setState = namedSet
-    const initialState = fn(namedSet, get, api)
-    if (!api.devtools) {
-      const savedSetState = api.setState
-      api.setState = <
-        K1 extends keyof S = keyof S,
-        K2 extends keyof S = K1,
-        K3 extends keyof S = K2,
-        K4 extends keyof S = K3
-      >(
-        state: PartialState<S, K1, K2, K3, K4>,
-        replace?: boolean
-      ) => {
-        const newState = api.getState()
-        if (state !== newState) {
-          savedSetState(state, replace)
-          if (state !== (newState as any)[DEVTOOLS] && api.devtools) {
-            api.devtools.send(api.devtools.prefix + 'setState', api.getState())
+    const setStateFromDevtools: SetState<S> = (state, replace) => {
+      isRecording = false
+      if (!replace) {
+        set(state)
+      } else {
+        let currentState = get()
+        for (let k in currentState) {
+          if (typeof currentState[k] === 'function') {
```

Comment:
I assume this is to recover non-serializable parameters. But, it feels a bit heuristic.
Can we give it up and just do `set(state)`, which is more predictable as the merging is the default behavior?

Reply:
For almost all operations we need replacement. For example `JUMP_STATE`:

Imagine the store initially is `{ foo: 0, someEffect: () => {} }` then user does `set({ bar: 0 })`. So in the eyes of the devtools the state first was `{ foo: 0 }` and now currently it is `{ foo: 0, bar: 0 }`.

So now when the users time travels to the previous state via the slider we'd receive the message `{ type: "DISPATCH", payload: { type: "JUMP_STATE" }, state: "{ 'foo': 0 }" }` and now if we merge the state ie `set({ foo: 0 })`, `bar` doesn't get removed. So the state should be `{ foo: 0, doEffect: () => {}` after time-travel but it's still the same ie `{ foo: 0, bar: 0, doEffect: () => {} }`

So now given we need to replace the state, we can't replace it directly as `set({ foo: 0 }, true)` would remove `doEffect` hence we do replace the state but we also copy the top-level functions to the new state so that they don't get removed.

I hope I've made myself clear. And yes indeed it's a heuristic, the idealistic solution would be to somehow serialize everything.

Reply:
I know what you mean and JUMP_STATE doesn't work nice for zustand.

There's no logic to assume there are top-level functions. It's just a convention to add action functions. So, I don't like using the assumption here. It should be more predictable than heuristics.

> to somehow serialize everything.

`Map<SomeSerializableIdentifier, UnserializableObject>`?

Another solution would be to keep the list of (nested) properties that are not serialized, and recover them. But, this is still less predictable, or uneasy.

So, I think it's better to educate to avoid using optional values. Instead of `set({ bar: 0 })`, do `set({ foo: null, bar: 0 })`.

Reply:
> `Map<SomeSerializableIdentifier, UnserializableObject>` ?

Kind of.

I agree that this approach is not the best but I realized we have problems with merging too (along with current approach): Imagine if the state was `{ posts: [] }` then the user did `set(s => ({ posts: [...s.posts, { title: "foo", delete }] })` so for devtools it's from `{ posts: [] }` to `{ posts: [{ title: "foo" }] }`. Now imagine if the user goes one step back, we do `set({ posts: [] })`, eveything good, then the users goes forward and we do `set({ posts: [{ title: "foo" }] })` and now the state no longer has the `delete` function.

So all approaches are only half-good, unless as I said we take the pain and figure out how to somehow serialize everything. I'm not sure if this is even possible, I guess the map approach would work. I think we should give that a try, or maybe pick an half-good approach (like merging) for now and keep his for next versions.

Wdyt? I'm open for everything.

Reply:
For this PR, I don't want to do fancy staff.
Let's either just merge or simply disable some features.
Let's revisit it after v4.

Reply:
Done we're now merging state